Source/Dafny/Resolver.cs

```diff
@@ -8983,6 +8992,21 @@ public class ResolveTypeReturn
           // resolve the whole thing
           ResolveConcreteUpdateStmt(s.Update, codeContext);
         }
+        if (s.Update is AssignOrReturnStmt)
+        {
+          var assignOrRet = (AssignOrReturnStmt)s.Update;
+          // resolve the LHS
+          Contract.Assert(assignOrRet.Lhss.Count == 1);
+          Contract.Assert(s.Locals.Count == 1);
+          var local = s.Locals[0];
+          var lhs = (IdentifierExpr)assignOrRet.Lhss[0];  // the LHS in this case will be an IdentifierExpr, because that's how the parser creates the VarDeclStmt
```

Comment:
Good to include cast-justifications like this in the comments. They explain why you expect the cast to hold (and shows you've given some thought to it).